design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,217 @@
+After the pod is scheduled, the now odd number of CPUs will collide with the CPU-Manager policy of using only full physical cores, and cause the kubelet to [reject](https://github.com/kubernetes/kubernetes/pull/101432) this pod.
+For example, let’s consider this VM spec:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+kubelet will emit following event:
+```
+SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
+```
+The following picture helps show how this is a problem
+<img src="smt-slignment-error-diagram.png">
+
+## Goals
+Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.
+
+## Non Goals
+Set the virt-launcher Pod scheduling to match a node with SMT aligned to its request.
+
+## Definition of Users
+This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.
+
+## User Stories
+* As a cluster admin, I want to enable VMs with an even number of CPUs and `isolatedEmulatorThread` enabled to run on my cluster by setting a cluster-wide knob on HCO.
+* As a Kubevirt VM owner, I should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
+
+## Repos
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+- [kubevirt/hyperconverged-cluster-operator](https://github.com/kubevirt/hyperconverged-cluster-operator)
+
+# Bug fix design proposal
+The bug fix proposal introduces an alpha annotation to the kubevirt CR: `kubevirt.io/EmulatorThreadCompleteToEvenParity`:
+```
+apiVersion: kubevirt.io/v1
+kind: KubeVirt
+metadata:
+  annotations:
+    kubevirt.io/EmulatorThreadCompleteToEvenParity:
+spec:
+  configuration:
+    developerConfiguration:
+      featureGates:
+      - EmulatorThreadCompleteToEvenParity
+```
+This annotation will be respected only when the `EmulatorThreadCompleteToEvenParity` featureGate is enabled.
+
+## Virt-API:
+
+### VMI mutator webhook
+When the following conditions are met:
+1. The feature gate is enabled on the Kubevirt CR
+2. Kubevirt CR has the annotation.
+3. VMI is configured with `isolateEmulatorThread: true`
+
+then the VMI mutator webhook will copy the annotation from the kubevirt CR to the VMI annotations. This is done in order to pass the information between the virt-controller, virt-handler and virt-launcher, without changing the API between the virt-handler and the virt-launcher.
+
+### VMI create-admitter validation webhook
+Should allow the VMI to have the new annotation only if the feature gate is enabled on the Kubevirt CR
```

Comment:
me & @vladikr agreed that for now, it is ok if the VMI is going to be admitted to the system. We will discuss guidance on future features and when this feature gets a formal API.

Reply:
So.. entirely remove this check from the create-admitter validation webhook, right?